rfcs/2023-09-19-18604-metrics-data-model-refactor.md

```diff
@@ -0,0 +1,466 @@
+rest represent the individual buckets or quantiles of the distribution.
+
+While these data types allow us to easily ingest Prometheus data, as well as emit it, problems arise
+when transforming these metrics to other metrics systems. In particular, many metrics systems have
+no concept at all of an "aggregated histogram". Perhaps the system only supports counters, gauges,
+and histograms (but aggregated server-side, so Vector only sends the raw samples) and so the
+aggregated metrics have to be decomposed... or perhaps the aggregated histograms can be transformed
+to something like [DDSketch](https://www.vldb.org/pvldb/vol12/p2195-masson.pdf). Whatever the
+mechanism, it's an additional metric type that must be supported individually for all metrics sinks.
+
+### Needless differentiation within metric types
+
+An example of this is the inherent "statistics" for a distribution metric.
+
+In practice, this leads to Vector users being forced to choose/consider the difference between a
+"histogram" and a "summary", when internally, there is no actual difference: a distribution carries
+raw sample values, and merely calculate summary statistics over those sample values. For metric
+sinks, those summary statistics are never used, as we simply deal with the raw sample values.
+
+Beyond the user experience pain, this is extra code that needs to be handled/dealt with internally,
+even when (again) it has no inherent value.
+
+## Proposal
+
+We would make the following changes to the metrics data model:
+
+- remove `MetricValue::AggregatedSummary` and `MetricValue::Set` entirely
+- refactor `MetricValue::Distribution` into an enum with a new `Raw` variant
+- merge `MetricValue::AggregatedHistogram` into `MetricValue::Distribution` as `FixedHistogram`
+- merge `MetricValue::Sketch` directly into `MetricValue::Distribution`, removing `MetricSketch` and
+  supporting DDSketch directly
+
+For a more visual representation, we would refactor `MetricValue` from this:
+
+```rust
+pub enum MetricValue {
+    Counter { value: f64 },
+    Gauge { value: f64 },
+    Set { values: BTreeSet<String> },
+    Distribution { samples: Vec<Sample>, statistic: StatisticKind },
+    AggregatedHistogram { buckets: Vec<Bucket>, count: u64, sum: f64 },
+    AggregatedSummary { quantiles: Vec<Quantile>, count: u64, sum: f64 },
+    Sketch { sketch: MetricSketch },
+}
+```
+
+to this:
+
+```rust
+pub enum Distribution {
+    Raw { samples: Vec<Sample> },
+    Histogram { buckets: Vec<Bucket>, count: u64, sum: f64 },
+    Sketch { sketch: AgentDDSketch },
+}
+
+pub enum MetricValue {
+    Counter { value: f64 },
+    Gauge { value: f64 },
+    Distribution { distribution: Distribution },
+}
+```
+
+### Removal of `AggregatedSummary`
+
+We would remove `AggregatedSummary` entirely, opting to handle them by decomposing them into
+individual metrics when received, and reassembling them (if need be) back into a single metric when
+shipping them out.
```

Comment:
It might be worth noting that this appears to only applies on the source side to the Prometheus sources, which already ship this type as individual metrics. Also, I am unclear how sinks will know to reassemble the collection.

Reply:
It doesn't actually ship them as individual metrics, though, at least not in the exposition format.

An individual aggregated summary metric looks something like:

```
# TYPE telemetry_requests_metrics_latency_microseconds summary
telemetry_requests_metrics_latency_microseconds{quantile="0.01"} 3102
telemetry_requests_metrics_latency_microseconds{quantile="0.05"} 3272
telemetry_requests_metrics_latency_microseconds{quantile="0.5"} 4773
telemetry_requests_metrics_latency_microseconds{quantile="0.9"} 9001
telemetry_requests_metrics_latency_microseconds{quantile="0.99"} 76656
telemetry_requests_metrics_latency_microseconds_sum 1.7560473e+07
telemetry_requests_metrics_latency_microseconds_count 2693
```

There is a line per individual _component_ of the metric, but it is a single metric based on the grouping inherent to the exposition format. One thing I can test is if it would be possible to emit the individual components (which would look like the following output below) and have Prometheus still function correctly:

```
# TYPE telemetry_requests_metrics_latency_microseconds counter
telemetry_requests_metrics_latency_microseconds{quantile="0.01"} 3102
telemetry_requests_metrics_latency_microseconds{quantile="0.05"} 3272
telemetry_requests_metrics_latency_microseconds{quantile="0.5"} 4773
telemetry_requests_metrics_latency_microseconds{quantile="0.9"} 9001
telemetry_requests_metrics_latency_microseconds{quantile="0.99"} 76656

# TYPE telemetry_requests_metrics_latency_microseconds_sum counter
telemetry_requests_metrics_latency_microseconds_sum 1.7560473e+07

# TYPE telemetry_requests_metrics_latency_microseconds_count counter
telemetry_requests_metrics_latency_microseconds_count 2693
```

This is perhaps an unspoken/unofficial way to represent an aggregated summary by its individual components that ultimately provides functional parity while being unconventional. If this _did_ work, it would neatly solve the issue of recomposing since we already group metrics by their series name to properly adhere to the exposition format.

Reply:
Testing this out locally, it doesn't actually appear to change anything about the behavior on the Prometheus side.

I had Prometheus set up to scrape aggregated summary metrics from unmodified version of Vector, and then a modified version of Vector scraping the unmodified version of Vector, doing the above conversion in the `prometheus_scrape` source, and sending them as-is to a `prometheus_exporter` sink.

Everything is fine on the Prometheus side, which as I type it and was doing this, I realized is precisely because there's no actual Prometheus query functions, etc, that deal with aggregated summaries, only histograms. So aggregated summaries are really just treated as their constituent parts i.e. they're a dumb of disparate series only sharing a common prefix and nothing else.

I'm going to update the RFC to reflect this, knowing now that we can just entirely ignore any reassembling on the sink side.

Reply:
Addressed in 951e806f9.